### PR TITLE
--authfile command line argument for image sign command.

### LINKED ISSUE #10866: podman image sign does not accept --authfile parameter
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind feature

**Description**

`podman image sign` does not accept the `authfile` parameter. To be consistent and to support those that can not use the default auth file location, need support for this flag.

**Steps to reproduce the issue:**

1. attempt to use the `--authfile` parameter for `podman image sign`

**Describe the results you received:**
`podman image sign` does not accept the `authfile` parameter.

**Describe the results you expected:**
`podman image sign` should accept the `authfile` parameter.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `podman version`:**

```
Version:      3.0.2-dev
API Version:  3.0.0
Go Version:   go1.15.13
Built:        Tue Jun  8 07:52:06 2021
OS/Arch:      linux/amd64
```

**Output of `podman info --debug`:**

```
host:
  arch: amd64
  buildahVersion: 1.19.8
  cgroupManager: systemd
  cgroupVersion: v1
  conmon:
    package: conmon-2.0.26-1.module+el8.4.0+10607+f4da7515.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.0.26, commit: b883692702312720058141f16b6002ab26ead2e7'
  cpus: 2
  distribution:
    distribution: '"rhel"'
    version: "8.4"
  eventLogger: file
  hostname: bastion.tssc.rht-set.com
  idMappings:
    gidmap: null
    uidmap: null
  kernel: 4.18.0-305.el8.x86_64
  linkmode: dynamic
  memFree: 2309537792
  memTotal: 3915423744
  ociRuntime:
    name: runc
    package: runc-1.0.0-73.rc93.module+el8.4.0+11311+9da8acfb.x86_64
    path: /usr/bin/runc
    version: |-
      runc version spec: 1.0.2-dev
      go: go1.15.13
      libseccomp: 2.5.1
  os: linux
  remoteSocket:
    path: /run/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_NET_RAW,CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: false
    seccompEnabled: true
    selinuxEnabled: true
  slirp4netns:
    executable: ""
    package: ""
    version: ""
  swapFree: 0
  swapTotal: 0
  uptime: 507h 13m 20.53s (Approximately 21.12 days)
registries:
  search:
  - registry.access.redhat.com
  - registry.redhat.io
  - docker.io
store:
  configFile: /etc/containers/storage.conf
  containerStore:
    number: 3
    paused: 0
    running: 0
    stopped: 3
  graphDriverName: overlay
  graphOptions:
    overlay.mountopt: nodev,metacopy=on
  graphRoot: /var/lib/containers/storage
  graphStatus:
    Backing Filesystem: xfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "true"
  imageStore:
    number: 270
  runRoot: /var/run/containers/storage
  volumePath: /var/lib/containers/storage/volumes
version:
  APIVersion: 3.0.0
  Built: 1623138726
  BuiltTime: Tue Jun  8 07:52:06 2021
  GitCommit: ""
  GoVersion: go1.15.13
  OsArch: linux/amd64
  Version: 3.0.2-dev
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
podman-3.0.1-7.module+el8.4.0+11311+9da8acfb.x86_64
```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide? (https://github.com/containers/podman/blob/master/troubleshooting.md)**

No

**Additional environment details (AWS, VirtualBox, physical, etc.):**
not environment specific.

### PR DESCRIPTION
**Pull request by jvanz**
Adds the --auth-file command line argument to allow users to define the signature file name.

Closes #10866

### CODE REVIEW

**Review by Luap99 — COMMENTED** [review]
The flag should be named `authfile` and not `auth-file` to be consistent with the other commands.

File: docs/source/markdown/podman-image-sign.1.md
Comment (mheon) [inline docs/source/markdown/podman-image-sign.1.md:41]: Can you make this more descriptive, please? This doesn't make it very clear what the flag does

**Comment by mheon** [comment]
@edsantiago Could I trouble you to take a peek at the tests added here, considering a new file / set of tests are being added?

**Review by edsantiago — CHANGES_REQUESTED** [review]
Do not merge. New tests need work. I will review later.

**Review by edsantiago — CHANGES_REQUESTED** [review]
Too difficult to write comments, it was easier to just make and test my rewrites. Could you please use [this 011-image.bats](https://github.com/containers/podman/files/6842858/011-image.bats.txt) instead? TL;DR (1) don't muck with user gpg files or current directory and don't use constant-name files/dirs in /tmp; (2) improve the signature file check

**Comment by edsantiago** [comment]
Also, heads-up to @lsm5 and @jnovy, this will require adding `Requires: gnupg` to the `-tests` subpackage in all specfiles.

**Comment by edsantiago** [comment]
@jvanz thanks for your PR! In addition to the comments above, please rebase to a more recent `main`, in order to get rid of the APIv2 test flake.

File: docs/source/markdown/podman-image-sign.1.md
Comment (TomSweeneyRedHat) [reply docs/source/markdown/podman-image-sign.1.md:41]: At the very least, you need a period at the end.

```suggestion
Override the default auth file of the signature.
```

Also, this should be `authfile`, not `auth-file`.  I think you can just grab the verbiage from http://docs.podman.io/en/latest/markdown/podman-pull.1.html#authfile-path or some of the other authfile options in Podman and put it in here too.

File: docs/source/markdown/podman-image-sign.1.md
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-image-sign.1.md:39]:
```suggestion
#### **--auth-file**=*file*
```

File: cmd/podman/images/sign.go
Comment (TomSweeneyRedHat) [inline cmd/podman/images/sign.go:52]: Please remove the dash here and anywhere that I might have missed.
```suggestion
	authFileFlagName := "authfile"
```

**Comment by vrothberg** [comment]
Friendly ping

File: cmd/podman/images/sign.go
Comment (Luap99) [inline cmd/podman/images/sign.go:53]: This should use `auth.GetDefaultAuthFile()`. Similar to this: https://github.com/containers/podman/blob/95ac8f180f26d7ed4cf2945aa206a156b7a169ca/cmd/podman/images/pull.go#L107-L109

**Comment by rhatdan** [comment]
LGTM
@Luap99 @vrothberg @TomSweeneyRedHat PTAL

File: docs/source/markdown/podman-image-sign.1.md
Comment (vrothberg) [inline docs/source/markdown/podman-image-sign.1.md:41]: It's not really the auth file **of** the signature, but the file used to look up credentials when contacting registries.

For the sake of consistency with all other commands with --authfile, please use the same description (e.g., https://github.com/containers/podman/blob/main/docs/source/markdown/podman-pull.1.md#--authfilepath).

**Review by vrothberg — COMMENTED** [review]
Let's wait for @edsantiago's review before merging.

File: test/system/011-image.bats
Comment (edsantiago) [inline test/system/011-image.bats:29]: skopeo tests include this magic code snippet, dating back several years:
```bash
    # gpg on f30 needs this, otherwise:
    #   gpg: agent_genkey failed: Inappropriate ioctl for device
    # ...but gpg on f29 (and, probably, Ubuntu) doesn't grok this
    GPGOPTS='--pinentry-mode loopback'
    if gpg --pinentry-mode asdf 2>&1 | grep -qi 'Invalid option'; then
        GPGOPTS=
    fi
```
It's possible that your use of `%no-ask-passphrase` solves this (I can't easily check right now). I mention it just to cover all bases.

File: test/system/011-image.bats
Comment (edsantiago) [inline test/system/011-image.bats:43]: Any reason you don't check the actual gpg output?
```diff
index 078b422c0..57763b635 100644
--- a/test/system/011-image.bats
+++ b/test/system/011-image.bats
@@ -40,7 +40,9 @@ function check_signature() {
     test -e "$dir/$1" || die "Missing signature file '$1'"

     # Confirm good signature
-    GNUPGHOME=$_GNUPGHOME_TMP gpg --verify "$dir/$1"
+    run env GNUPGHOME=$_GNUPGHOME_TMP gpg --verify "$dir/$sigfile"
+    is "$output" ".*Good signature from .Foo.*<foo@bar.com>" \
+       "gpg --verify $sigfile"
 }


```
(Requires adding `local sigfile=$1`, which I would encourage you to do when `$1` is used in more than one or two places).

File: test/system/011-image.bats
Comment (edsantiago) [inline test/system/011-image.bats:21]: As a general rule I like to make these verifiable strings random, to prevent errors such as tests reusing a GNUPGHOME dir or config. In this case I think the chances of that are slim, so I'm 80% OK with leaving it as-is.

File: docs/source/markdown/podman-image-sign.1.md
Comment (edsantiago) [inline docs/source/markdown/podman-image-sign.1.md:43]: Oops, one more: I find this wording confusing. First, the `, which` in line 41 has an ambiguous referent and could be rewritten to make it clearer. Second, I really don't understand the chain of precedence. Is the default line auth.json, or config.json, or what? I would suggest looking at this with fresh eyes.

**Review by edsantiago — COMMENTED** [review]
LGTM, with a few suggestions, and one more suggestion: please add `Requires: gnupg` to the `%package tests` section in `contrib/spec/podman.spec.in`.

And, BIG HEADS UP TO @lsm5 and @jnovy: the same (adding gnupg) will need to be done to all real specfiles!

I am OOTO all of Tuesday, and cannot follow up until Tuesday night or Wednesday morning.

**Comment by vrothberg** [comment]
friendly ping

File: docs/source/markdown/podman-image-sign.1.md
Comment (rhatdan) [reply docs/source/markdown/podman-image-sign.1.md:43]: Removed to match podman create.

**Pull request closed without merging**
